<a href="https://colab.research.google.com/github/APWright/CSC477-Fall2025/blob/main/Assignment-2/Assignment_2_template.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Assignment 2: Data Wrangling


  - To complete this excercise start by saving a copy of this template to google drive using the "Copy to Drive" button

  - In your copy, complete the notebook directly in place at the sections labeled:
  ```python
  # Your code here
  ```

  - Once you have finished the exercise, click the Share button at the top of the notebook. Provide anyone with the link view access, then copy the link and submit it in Canvas.
  
  - Additionally, under File>Download save the notebook file and submit it in Canvas as well.

In this lab we will be using the dataframe library *polars* and the data visualization library *vega-altiar*. If you are already familiar with libraries such as pandas a lot will be transferable, however altair does work meaningfully differently than other common plotting libraries like matplotlib. First we should start by importing the required libraries and downloading the required datasets by running the following cell.

In [1]:
# @title Import libraries and setup environment
%%capture
%pip install vegafusion vl-convert-python

!git clone https://github.com/APWright/CSC477-Fall2025.git
%cd  CSC477-Fall2025/Assignment-2/


import polars as pl
import altair as alt
import vega_datasets

zipcodes = vega_datasets.data.zipcodes().to_dict('records')
population = pl.read_csv('population.csv')
population_wide = pl.read_csv('population_wide.csv')
pop_untidy = pl.read_csv('pop.csv',infer_schema=False)
gdp_untidy = pl.read_csv('gdp_pcap.csv',infer_schema=False)
emissions_untidy = pl.read_csv('co2_pcap_cons.csv',infer_schema=False)

alt.data_transformers.enable("vegafusion")

# Part 1: Data wrangling without Dataframes

For this exercise, we'll use a dataset of US zipcodes.
The dataset is available in the zipcodes array, which is an array of objects. As an example, here's what the first object looks like:

In [2]:
zipcodes[0]

{'zip_code': '00501',
 'latitude': 40.922326,
 'longitude': -72.637078,
 'city': 'Holtsville',
 'state': 'NY',
 'county': 'Suffolk'}



For this part we are working through how data wranging and analysis might work without the use of dedicated DataFrame apis. Therefore for the following cells only use built-in python language features such as Using [`map`](https://docs.python.org/3/library/functions.html#map), [`filter`](https://docs.python.org/3/library/functions.html#filter), and [list comprehensions](https://www.w3schools.com/python/python_lists_comprehension.asp).

## 1.1: Find the zipcodes associated with the city of San Luis Obispo.
___
Cell should output a list of zipcode records.






In [ ]:
answer_1_1 = ...

# Your code here

answer_1_1

## 1.2: How many zip codes are associated with the state of California?

---



Cell should output a number.






In [ ]:
answer_1_2 =  ...

# Your code here

answer_1_2

## 1.3: Derived Fields
___

Often in visualization, we want to display derived fields, rather than only the fields in the original dataset.

We would like to compute, for each record, the first character of each zipcode. For example, for the zipcode object associated with 93405, we want the following record:


```python
{
  'zip_code': '93401',
  'latitude': 35.265573,
  'longitude': -120.62122,
  'city': 'San Luis Obispo',
  'state': 'CA',
  'county': 'San Luis Obispo'
  'first_digit' : 9 # <--- New field added!
}
  ```

Produce a new list of records that includes a new field called `first_digit` that contains this value.

Expected answer is a list of zipcode objects, with each object now also having a `first_digit` key.

In [ ]:
answer_1_3 = ...

# Your code here

answer_1_3 = list(map(add_first_digit, zipcodes))

## 1.4: Find the maximum zipcode in the US.
___
[Is this a meaningful value?](https://en.wikipedia.org/wiki/ZIP_Code#Geographic_hierarchy)

In [ ]:
answer_1_4 = ...

# Your code here

answer_1_4

# Part 2: Data wrangling with Polars

We can do quite a bit with built-in data strucutres like lists, arrays, dicts, and built-in functions like map, filter, and list comprehensions. However, other common data wrangling operations are unavailable through arrays (like group by operations), or would be inefficient when working with large datasets.

That's where [Polars](https://docs.pola.rs/) comes in. Polars is an open source DataFrame library much like (and often interoperable with) [Pandas](https://pandas.pydata.org/) but written in Rust for better performance.

***STOP BEFORE YOU CONTINUE***

Peruse the [polars user guide ](https://docs.pola.rs/user-guide/getting-started/) before moving on with this section, and keep it as well as the [API Reference](https://docs.pola.rs/api/python/stable/reference/index.html) handy. This lab assignment is intentionally light on direct instruction, partly because the linked resources do it better, and partly because in an upper division course you should be used to reading documentation.



Now we can get started and import our zipcode data into a dataframe. The following exercises will ask you to perform some data tasks on the dataframe `df`.


In [3]:
import polars as pl
df = pl.DataFrame(zipcodes)
df

shape: (42_049, 6)
┌──────────┬───────────┬─────────────┬─────────────┬───────┬────────────────────────┐
│ zip_code ┆ latitude  ┆ longitude   ┆ city        ┆ state ┆ county                 │
│ ---      ┆ ---       ┆ ---         ┆ ---         ┆ ---   ┆ ---                    │
│ str      ┆ f64       ┆ f64         ┆ str         ┆ str   ┆ str                    │
╞══════════╪═══════════╪═════════════╪═════════════╪═══════╪════════════════════════╡
│ 00501    ┆ 40.922326 ┆ -72.637078  ┆ Holtsville  ┆ NY    ┆ Suffolk                │
│ 00544    ┆ 40.922326 ┆ -72.637078  ┆ Holtsville  ┆ NY    ┆ Suffolk                │
│ 00601    ┆ 18.165273 ┆ -66.722583  ┆ Adjuntas    ┆ PR    ┆ Adjuntas               │
│ 00602    ┆ 18.393103 ┆ -67.180953  ┆ Aguada      ┆ PR    ┆ Aguada                 │
│ 00603    ┆ 18.455913 ┆ -67.14578   ┆ Aguadilla   ┆ PR    ┆ Aguadilla              │
│ …        ┆ …         ┆ …           ┆ …           ┆ …     ┆ …                      │
│ 99926    ┆ 55.094325 ┆ -131.566827 ┆ Metlakatla  ┆ AK    ┆ Prince Wales Ketchikan │
│ 99927    ┆ 55.517921 ┆ -132.003244 ┆ Point Baker ┆ AK    ┆ Prince Wales Ketchikan │
│ 99928    ┆ 55.395359 ┆ -131.67537  ┆ Ward Cove   ┆ AK    ┆ Ketchikan Gateway      │
│ 99929    ┆ 56.449893 ┆ -132.364407 ┆ Wrangell    ┆ AK    ┆ Wrangell Petersburg    │
│ 99950    ┆ 55.542007 ┆ -131.432682 ┆ Ketchikan   ┆ AK    ┆ Ketchikan Gateway      │
└──────────┴───────────┴─────────────┴─────────────┴───────┴────────────────────────┘


---



## 2.1: How many zip codes are there in the US? (Or, how many rows are in your data table?)
___

In [ ]:
answer_2_1 = ...

# Your Code Here

answer_2_1

## 2.2 How many counties does each state have?
___
 Hint: you will want to look at df.groupby (to group by state), agg (to summarize per state), and n_unique(to find the number of unique county names).

  You'll know your answer is right if you see that California has 58 counties.

In [ ]:
answer_2_2 = ...

# Your Code Here

answer_2_2

## 2.3 Derive a column
___
Same as you did with the array in Part 1, we'll now derive a column for our data table. Add a new last_digit column to the data table that is simply the last digit of the zip codes, and cast it as a 8 bit integer (`pl.Int8`). Instead of simply viewing the result, save it in a new variable.

It will be useful to have read the documentation on [Expressions](https://docs.pola.rs/user-guide/concepts/expressions-and-contexts/), [Strings](https://docs.pola.rs/user-guide/expressions/strings/), and [Casting](https://docs.pola.rs/user-guide/expressions/casting/).


In [ ]:
answer_2_3 = ...

# Your Code Here

answer_2_3

# A sneak peak of what's to come

So how do we use this data in visualization? For most of this course, we're going to use a visualization library called [Vega-Altair](https://altair-viz.github.io/index.html) that allows us to create powerful interactive graphics using a relatively simple declarative syntax. Polars provides built in support for Altair.

Run the below cell to see a plot of US zip codes. Expand the source code for the cell below and see the comments, along with some nudges to experiment.

In [ ]:
# @title US Zip Codes

import altair as alt
import vega_datasets

# Read in polygons from topojson
states = alt.topo_feature(vega_datasets.data.us_10m.url, feature='states')

#US states background
background = alt.Chart(states).mark_geoshape(
    fill='lightgray',
    stroke='white'
).properties(
    width=500,
    height=300
).project('albersUsa')
"""
The whole chart is projected (roughly, "scaled") using the Albers USA projection.
It takes certain liberties with the locations of Alaska and Hawaii to present the
map in a confined space.
Comment this out to see what happens without it.
Also try different projections, like "albers", "equalEarth", or "mercator".
"""

# plot zip positions on background
points = alt.Chart(df).mark_circle(
    size=10,
    opacity=0.6,
).encode(
    longitude='longitude:Q',
    latitude='latitude:Q',
    tooltip='county',
    color='first_digit'
)

#Layer points and background
chart = alt.layer(
    background, points
).properties(
    width=500,
    height=500,
    padding=10  # Padding ensures labels fit
).interactive()

chart

# Part 3: Into the Tidyverse

In class we talked about Tidy data. A tidy dataset has three properties:

1.  Each variable has its own column.
1.    Each observation has its own row.
1.    Each value has its own cell.


Consider this table of US population counts from 1850 to 2000. It is tidy, since each observation is a row, and each variable is a column.

The **age** column depicts an **age** *range*, i.e., when age is 0, it means people aged 0–4. When **age** is 5, it means people aged 5–9.



In [4]:
population

shape: (570, 4)
┌──────┬─────┬────────┬─────────┐
│ year ┆ age ┆ sex    ┆ people  │
│ ---  ┆ --- ┆ ---    ┆ ---     │
│ i64  ┆ i64 ┆ str    ┆ i64     │
╞══════╪═════╪════════╪═════════╡
│ 1850 ┆ 0   ┆ male   ┆ 1483789 │
│ 1850 ┆ 0   ┆ female ┆ 1450376 │
│ 1850 ┆ 5   ┆ male   ┆ 1411067 │
│ 1850 ┆ 5   ┆ female ┆ 1359668 │
│ 1850 ┆ 10  ┆ male   ┆ 1260099 │
│ …    ┆ …   ┆ …      ┆ …       │
│ 2000 ┆ 80  ┆ female ┆ 3221898 │
│ 2000 ┆ 85  ┆ male   ┆ 970357  │
│ 2000 ┆ 85  ┆ female ┆ 1981156 │
│ 2000 ┆ 90  ┆ male   ┆ 336303  │
│ 2000 ┆ 90  ┆ female ┆ 1064581 │
└──────┴─────┴────────┴─────────┘

The same data could be represented in the following decidedly *un*-tidy table.

In [5]:
population_wide

shape: (30, 21)
┌────────┬──────┬─────────┬──────────┬───┬─────────┬─────────┬─────────┬─────────┐
│ sex    ┆ year ┆ 0       ┆ 5        ┆ … ┆ 75      ┆ 80      ┆ 85      ┆ 90      │
│ ---    ┆ ---  ┆ ---     ┆ ---      ┆   ┆ ---     ┆ ---     ┆ ---     ┆ ---     │
│ str    ┆ i64  ┆ i64     ┆ i64      ┆   ┆ i64     ┆ i64     ┆ i64     ┆ i64     │
╞════════╪══════╪═════════╪══════════╪═══╪═════════╪═════════╪═════════╪═════════╡
│ male   ┆ 1850 ┆ 1483789 ┆ 1411067  ┆ … ┆ 40834   ┆ 23449   ┆ 8186    ┆ 5259    │
│ female ┆ 1850 ┆ 1450376 ┆ 1359668  ┆ … ┆ 40229   ┆ 22949   ┆ 10511   ┆ 6569    │
│ male   ┆ 1860 ┆ 2120846 ┆ 1804467  ┆ … ┆ 56699   ┆ 29007   ┆ 10434   ┆ 7232    │
│ female ┆ 1860 ┆ 2092162 ┆ 1778772  ┆ … ┆ 57597   ┆ 29506   ┆ 14053   ┆ 6622    │
│ male   ┆ 1870 ┆ 2800083 ┆ 2428469  ┆ … ┆ 86929   ┆ 47427   ┆ 15891   ┆ 8649    │
│ …      ┆ …    ┆ …       ┆ …        ┆ … ┆ …       ┆ …       ┆ …       ┆ …       │
│ female ┆ 1980 ┆ 8081854 ┆ 8275881  ┆ … ┆ 2951759 ┆ 1919292 ┆ 1023115 ┆ 499046  │
│ male   ┆ 1990 ┆ 9307465 ┆ 9274732  ┆ … ┆ 2344932 ┆ 1342996 ┆ 588790  ┆ 238459  │
│ female ┆ 1990 ┆ 8894007 ┆ 8799955  ┆ … ┆ 3723980 ┆ 2545730 ┆ 1419494 ┆ 745146  │
│ male   ┆ 2000 ┆ 9735380 ┆ 10552146 ┆ … ┆ 2912655 ┆ 1902638 ┆ 970357  ┆ 336303  │
│ female ┆ 2000 ┆ 9310714 ┆ 10069564 ┆ … ┆ 4355644 ┆ 3221898 ┆ 1981156 ┆ 1064581 │
└────────┴──────┴─────────┴──────────┴───┴─────────┴─────────┴─────────┴─────────┘

Instead of having the different age groups be *values* in the age column, the table spreads them across *multiple rows*. This table can probably be made sense of by a human—for example, in 1850, there were 1,483,789 males in the US aged 0–4.

However, it really doesn't lend itself to useful programmatic or visual analyses.

## 3.1: Reshaping Dataframes
___
Polars provides the `pivot` and `unpivot` functions to help with reshaping data tables. Using the linked documentation as a guide, complete the following steps.

 Return a table that looks identical to `population_wide`.

 *Hint*: You don't have to type out all the column names in an array, and you can check with `.equals` method fo dataframes

In [ ]:
answer_3_1 = ...

# Your code here

answer_3_1

## 3.2: Folding, and merging multiple datasets
___
Consider the following three SUPER UNTIDY tables from the [Gapminder](https://www.gapminder.org/data/) project.

First, we have data about countries' per-capita CO2 emissions (in tonnes). The table is very wide...



In [6]:
emissions_untidy

shape: (194, 224)
┌──────────────┬───────┬───────┬───────┬───┬───────┬───────┬───────┬───────┐
│ country      ┆ 1800  ┆ 1801  ┆ 1802  ┆ … ┆ 2019  ┆ 2020  ┆ 2021  ┆ 2022  │
│ ---          ┆ ---   ┆ ---   ┆ ---   ┆   ┆ ---   ┆ ---   ┆ ---   ┆ ---   │
│ str          ┆ str   ┆ str   ┆ str   ┆   ┆ str   ┆ str   ┆ str   ┆ str   │
╞══════════════╪═══════╪═══════╪═══════╪═══╪═══════╪═══════╪═══════╪═══════╡
│ Afghanistan  ┆ 0.001 ┆ 0.001 ┆ 0.001 ┆ … ┆ 0.261 ┆ 0.261 ┆ 0.279 ┆ 0.284 │
│ Angola       ┆ 0.009 ┆ 0.009 ┆ 0.009 ┆ … ┆ 1.15  ┆ 1.12  ┆ 1.2   ┆ 1.23  │
│ Albania      ┆ 0.001 ┆ 0.001 ┆ 0.001 ┆ … ┆ 2.05  ┆ 2     ┆ 2.12  ┆ 2.1   │
│ Andorra      ┆ 0.333 ┆ 0.335 ┆ 0.337 ┆ … ┆ 6.06  ┆ 5.63  ┆ 5.97  ┆ 5.91  │
│ UAE          ┆ 0.063 ┆ 0.063 ┆ 0.064 ┆ … ┆ 21.2  ┆ 19.7  ┆ 20.7  ┆ 21.1  │
│ …            ┆ …     ┆ …     ┆ …     ┆ … ┆ …     ┆ …     ┆ …     ┆ …     │
│ Samoa        ┆ 0.002 ┆ 0.002 ┆ 0.002 ┆ … ┆ 1.37  ┆ 1.31  ┆ 1.4   ┆ 1.43  │
│ Yemen        ┆ 0.002 ┆ 0.002 ┆ 0.002 ┆ … ┆ 0.365 ┆ 0.362 ┆ 0.387 ┆ 0.395 │
│ South Africa ┆ 0.003 ┆ 0.003 ┆ 0.004 ┆ … ┆ 5.68  ┆ 5.11  ┆ 5.08  ┆ 5.18  │
│ Zambia       ┆ 0.255 ┆ 0.256 ┆ 0.257 ┆ … ┆ 0.487 ┆ 0.388 ┆ 0.416 ┆ 0.424 │
│ Zimbabwe     ┆ 0.059 ┆ 0.06  ┆ 0.06  ┆ … ┆ 0.783 ┆ 0.737 ┆ 0.789 ┆ 0.804 │
└──────────────┴───────┴───────┴───────┴───┴───────┴───────┴───────┴───────┘

Next, we have data about countries' populations and projected populations from 1800–2100. (Not to be confused with the US population dataset we played with in Part 3.1.)

In [7]:
pop_untidy

shape: (197, 302)
┌──────────────┬───────┬───────┬───────┬───┬───────┬───────┬───────┬───────┐
│ country      ┆ 1800  ┆ 1801  ┆ 1802  ┆ … ┆ 2097  ┆ 2098  ┆ 2099  ┆ 2100  │
│ ---          ┆ ---   ┆ ---   ┆ ---   ┆   ┆ ---   ┆ ---   ┆ ---   ┆ ---   │
│ str          ┆ str   ┆ str   ┆ str   ┆   ┆ str   ┆ str   ┆ str   ┆ str   │
╞══════════════╪═══════╪═══════╪═══════╪═══╪═══════╪═══════╪═══════╪═══════╡
│ Afghanistan  ┆ 3.28M ┆ 3.28M ┆ 3.28M ┆ … ┆ 128M  ┆ 129M  ┆ 130M  ┆ 130M  │
│ Angola       ┆ 1.57M ┆ 1.57M ┆ 1.57M ┆ … ┆ 147M  ┆ 148M  ┆ 149M  ┆ 150M  │
│ Albania      ┆ 400k  ┆ 402k  ┆ 404k  ┆ … ┆ 1.23M ┆ 1.22M ┆ 1.2M  ┆ 1.18M │
│ Andorra      ┆ 2650  ┆ 2650  ┆ 2650  ┆ … ┆ 49k   ┆ 48.4k ┆ 47.8k ┆ 47.2k │
│ UAE          ┆ 40.2k ┆ 40.2k ┆ 40.2k ┆ … ┆ 25.4M ┆ 25.7M ┆ 25.9M ┆ 26.1M │
│ …            ┆ …     ┆ …     ┆ …     ┆ … ┆ …     ┆ …     ┆ …     ┆ …     │
│ Samoa        ┆ 47.3k ┆ 47.3k ┆ 47.3k ┆ … ┆ 380k  ┆ 381k  ┆ 382k  ┆ 384k  │
│ Yemen        ┆ 2.59M ┆ 2.59M ┆ 2.59M ┆ … ┆ 109M  ┆ 109M  ┆ 110M  ┆ 110M  │
│ South Africa ┆ 1.45M ┆ 1.45M ┆ 1.46M ┆ … ┆ 93.7M ┆ 93.9M ┆ 94.1M ┆ 94.3M │
│ Zambia       ┆ 747k  ┆ 758k  ┆ 770k  ┆ … ┆ 63.4M ┆ 63.8M ┆ 64.1M ┆ 64.5M │
│ Zimbabwe     ┆ 1.09M ┆ 1.09M ┆ 1.09M ┆ … ┆ 36.9M ┆ 37M   ┆ 37.1M ┆ 37.2M │
└──────────────┴───────┴───────┴───────┴───┴───────┴───────┴───────┴───────┘

Finally, we have data (and projections) about the countries' GDP per capita (in inflation adjusted 2017 USD) from 1800–2100.

In [8]:
gdp_untidy

shape: (195, 302)
┌──────────────┬──────┬──────┬──────┬───┬───────┬───────┬───────┬───────┐
│ country      ┆ 1800 ┆ 1801 ┆ 1802 ┆ … ┆ 2097  ┆ 2098  ┆ 2099  ┆ 2100  │
│ ---          ┆ ---  ┆ ---  ┆ ---  ┆   ┆ ---   ┆ ---   ┆ ---   ┆ ---   │
│ str          ┆ str  ┆ str  ┆ str  ┆   ┆ str   ┆ str   ┆ str   ┆ str   │
╞══════════════╪══════╪══════╪══════╪═══╪═══════╪═══════╪═══════╪═══════╡
│ Afghanistan  ┆ 481  ┆ 481  ┆ 481  ┆ … ┆ 5380  ┆ 5510  ┆ 5640  ┆ 5780  │
│ Angola       ┆ 373  ┆ 374  ┆ 376  ┆ … ┆ 27.7k ┆ 28.2k ┆ 28.8k ┆ 29.3k │
│ Albania      ┆ 469  ┆ 471  ┆ 472  ┆ … ┆ 58k   ┆ 58.6k ┆ 59.2k ┆ 59.8k │
│ Andorra      ┆ 1370 ┆ 1370 ┆ 1370 ┆ … ┆ 81.5k ┆ 81.8k ┆ 82k   ┆ 82.3k │
│ UAE          ┆ 1140 ┆ 1150 ┆ 1150 ┆ … ┆ 92.8k ┆ 92.8k ┆ 92.8k ┆ 92.9k │
│ …            ┆ …    ┆ …    ┆ …    ┆ … ┆ …     ┆ …     ┆ …     ┆ …     │
│ Samoa        ┆ 1600 ┆ 1600 ┆ 1600 ┆ … ┆ 27k   ┆ 27.6k ┆ 28.1k ┆ 28.6k │
│ Yemen        ┆ 1010 ┆ 1010 ┆ 1020 ┆ … ┆ 7120  ┆ 7290  ┆ 7470  ┆ 7650  │
│ South Africa ┆ 1750 ┆ 1730 ┆ 1710 ┆ … ┆ 48.2k ┆ 48.8k ┆ 49.5k ┆ 50.1k │
│ Zambia       ┆ 533  ┆ 535  ┆ 536  ┆ … ┆ 19k   ┆ 19.4k ┆ 19.9k ┆ 20.3k │
│ Zimbabwe     ┆ 919  ┆ 920  ┆ 921  ┆ … ┆ 11.4k ┆ 11.6k ┆ 11.9k ┆ 12.2k │
└──────────────┴──────┴──────┴──────┴───┴───────┴───────┴───────┴───────┘

Our goal in this part is to all of this data into one tidy dataset. That is, we want a table that looks something like this.



| Country     | Year | Emissions | GDP | Population |
|-------------|------|-----------|-----|-------|
| Afghanistan | 1800 | 0.001 | 481 | 3.28M |
| Afghanistan | 1801 | 0.001 | 481 | 3.28M |
| Afghanistan | 1803 | 0.001 | 481 | 3.28M |
| Angola      | 1800 | 0.009 | 373 | 1.57M |
| Angola      | 1801 | 0.009 | 374 | 1.57M |
| Angola      | 1802 | 0.009 | 376 | 1.57M |
|	and |	so 	| on |	... |


Where each row is an observation (a specific country's values in a specific year) and each column is an attribute.

In order to do this you will need to first tidy up the individual tables of each data source. Then you will need to combine the individual tables together into one tidy dataframe. Consider the [`join`](https://docs.pola.rs/user-guide/transformations/joins/) operation, and make sure to be careful about the kind of join used, as not all of the tables have data for all of the same years. For example, what if one table had no CO2 data for a given year and country? Should that year be omitted entirely from the merged table? Or should those cells be given empty values?

Similarly, what if the columns were named different things, like year in one table and emissionYear in another table? The various join methods provide options for handling these cases. Please give them a brief look in case you encounter these situations in future assignments.

In [ ]:
answer_3_2 = ...

# Your code here

answer_3_2

## 3.3: Cleaning and Wrangling Data
___

Finally, you'll notice that the numerical fields don't really have numbers in them. The original data source (Gapminder) has provided the bigger numbers as strings, i.e., instead of the number 3280000, the table contains the value "3.28M".

This makes things easy for a human to read, but really isn't going to work for visualization.




The work of transforming, processing, and cleaning data to make it work for visualization and analysis is generally called *data wrangling* and in most data science and data engineering work takes up a shockingly large percentage of your actual time and effort. Polars provides many powerful and effeicent ways to do much of this work, in particular, ensure that you have fully understood Polars [expressions and contexts](https://docs.pola.rs/user-guide/concepts/expressions-and-contexts/).


In the function below, define a polars **expression** that when run in a `with_columns` context will transform a series of strings such as "3.28M" into a series of numbers of a given datatype (either floats or integers).

In [ ]:
def parse_number_shorthand(series_name, dtype):
  answer_3_3 = ...

  # Your code here

  return answer_3_3

In [ ]:
gapminder_tidy = answer_3_2.with_columns(
    parse_number_shorthand('year',pl.Int32),
    parse_number_shorthand('population',pl.Int32),
    parse_number_shorthand('gdp_per_capita',pl.Int32),
    parse_number_shorthand('emissions_per_capita',pl.Float32),
)
gapminder_tidy

Output should look somthing like this


| country	| year	| population| 	emissions_per_capita	|gdp_per_capita|
|--|--|--|--|--|
|str|	i32	|i32	|f32	|i32|
|--|--|--|--|--|
|"Afghanistan"|	1800	|3280000|	0.001	|481|
|"Angola"|	1800|	1570000|	0.009|	373|
|"Albania"|	1800|	400000|	0.001|	469|
|"Andorra"|1800|	2650|	0.333	|1370|
|"UAE"|	1800|	40200|	0.063|	1140|
|...|...|...|...|...|

## 3.4: Summarizing Data
___

Now that we have wrangled our data into a useable form we can start to do some analysis and data summarization. Here calculate the total cumulative overall (**not** per capita) CO2 emissions from the USA from 1800-2022.

In [ ]:
answer_3_4 = ...

# Your code here

answer_3_4

# Using Tidy Data for Interactive Visualization

At last now we have built a dataframe that can be used to asnwer questions about our data, and can be integrated with interactive data visualization libraries like Altair. Run the cell below for an example of a reporoduction of the [famous gapminder animation](https://www.youtube.com/watch?v=hVimVzgtD6w), where you can use the slider to progress through the years, hover over circles to see the countries they represent, and zoom in to find interesting patters. I recomend you spend some time going over the source code as well, and play around with the choices I made in order to start learning Altair as it will be the primary python based library we will be using.

In [ ]:
# @title Gapminder Interactive Animation
year_slider = alt.binding_range(min=1950,max=2022, name='Year')
selector = alt.selection_point(name='year_select',fields=['year'], bind=year_slider,value=1950)

scatter = alt.Chart(gapminder_tidy).mark_circle(opacity=0.80).encode(
        x=(alt.X('gdp_per_capita')
              .sort(field= 'population', order= 'descending')
              .scale(type="log",domain=(100, 1000000))
              .axis(grid=False)
              .title('GDP per capita, log scale')
          ),
        y=(alt.Y('emissions_per_capita')
              .sort(field= 'population', order= 'descending')
              .scale(type="log",domain=(0.001, 100))
              .axis(grid=False)
              .title('CO2 emissions (tonnes per capita), log scale')
          ),
        color=(alt.Color('country',legend=None)
              .sort(field= 'country', order= 'descending')
          ),
        size=(alt.Size('population',legend=None)
              .scale(range=(10, 10000))
          ),
        tooltip=['country','population','emissions_per_capita']
).add_params(
    selector
).transform_filter(
    selector
)

background_year = alt.Chart(gapminder_tidy).mark_text(
    baseline='middle',
    fontSize=75,
    opacity=0.1
).encode(
    text='year:O'
).transform_filter(
    selector
).transform_aggregate(
    year='max(year)'
)


chart = alt.layer(
    scatter, background_year
).properties(
    width=500,
    height=500,
    padding=10  # Padding ensures labels fit
).configure_axis(
    labelFontSize=12,
    titleFontSize=12
).interactive()

chart

